src/Unsafe/Linear.hs

```diff
@@ -20,42 +27,188 @@
-toLinear2 = coerce
+toLinear2 f = case unsafeEqualityProof @'(p,q) @'( 'One, 'One) of
+  UnsafeRefl -> f
 
 -- | Like 'toLinear' but for three-argument functions
 toLinear3
   :: forall (r1 :: RuntimeRep) (r2 :: RuntimeRep)
      (r3 :: RuntimeRep) (r4 :: RuntimeRep)
      (a :: TYPE r1) (b :: TYPE r2) (c :: TYPE r3) (d :: TYPE r4) p q r.
      (a %p-> b %q-> c %r-> d) %1-> (a %1-> b %1-> c %1-> d)
-toLinear3 = coerce
+toLinear3 f = case unsafeEqualityProof @'(p,q,r) @'( 'One, 'One, 'One) of
+  UnsafeRefl -> f
+
+
+-- | @toLinearN@ subsumes the functionality of 'toLinear1', 'toLinear2', and
+-- 'toLinear3'. In particular, @toLinearN \@n@ unsafely changes the
+-- multiplicities of the first @n@ arrows from any multiplicity to any
+-- other multiplicity.
+--
+-- === Examples
+--
+-- @
+-- toLinearN \@2 :: (a -> b -> Int) -> a %1-> b %1-> Int
+-- 'toLinear3' = toLinearN \@3
+-- @
+toLinearN :: forall n f g m. ToLinearN n f g => f %m-> g
+toLinearN f = case unsafeLinearityProofN @n @f @g of
+  UnsafeRefl -> f
+
+-- | @ToLinearN n f g@ means that @f@ and @g@ are the same with the
+-- possible exception of the multiplicities of the first @n@ arrows.
+type ToLinearN :: forall {rep :: RuntimeRep}. Nat -> TYPE rep -> TYPE rep -> Constraint
+class ToLinearN n f g where
+  -- | @unsafeLinearityProofN \@n \@f \@g @ is a fake proof that @f@ and @g@ are
+  -- the same with the possible exception of the multiplicities of the first
+  -- @n@ arrows.
+  unsafeLinearityProofN :: UnsafeEquality f g
+instance (ToLinearN' ni f g, UnifyL ni f g, UnifyL ni g f, ni ~ ToINat n) => ToLinearN n f g where
+  unsafeLinearityProofN = prf @ni
+
+-- | @toLinearL@ is similar to 'toLinearN', but it offers
+-- finer-grained control. The user supplies a type argument
+-- showing what they expect each of the arrows to be originally,
+-- and what each changes to. These may be polymorphic, and wildcards
+-- will work when they can be inferred.
+--
+-- === Examples
+--
+-- @
+-- toLinearL \@'[ 'Many :=> 'One, 'Many :=> 'One]
+--   :: (a -> b -> c) -> a %1-> b %1-> c
+-- 'toLinear3' = toLinearL \@'[ _ :=> 'One, _ :=> 'One, _ :=> 'One]
+-- @
+toLinearL :: forall (l :: [Change]) f g m. ToLinearL l f g => f %m-> g
+toLinearL f = case unsafeLinearityProofL @l @f @g of
+  UnsafeRefl -> f
+
+-- | @ToLinearL l f g@ means that @f@ and @g@ are the same with the
+-- possible exception of the multiplicities of some of the first arrows,
+-- as described by the list @l@.
+type ToLinearL :: forall {rep :: RuntimeRep}. [Change] -> TYPE rep -> TYPE rep -> Constraint
+class ToLinearL l f g where
+  -- | @unsafeLinearityProofL \@l \@f \@g @ is a fake proof that @f@ and @g@ are
+  -- the same with the possible exception of the multiplicities of the first
+  -- arrows given by the list @l@.
+  unsafeLinearityProofL :: UnsafeEquality f g
+instance ( ToLinearN' ni f g, UnifyL ni f g, UnifyL ni g f
+         , ni ~ Length l, UseChanges l f g) => ToLinearL l f g where
+  unsafeLinearityProofL = prf @ni
```

Comment:
Is this really necessary? This can be achieved with a `ToLinearN` and a type annotation. I think that I'm essentially questioning the need for exporting an operator from this `Unsafe` module.